src/dmd/must_use.d

```diff
@@ -0,0 +1,203 @@
+                    Id.udaMustUse.toChars(), sym.kind());
+                sym.errors = true;
+            }
+        }
+    }
+}
+
+/**
+ * Returns true if the given expression is an assignment, either simple (a = b)
+ * or compound (a += b, etc).
+ */
+private bool isAssignment(Expression e)
+{
+    if (e.isAssignExp || e.isBinAssignExp)
+        return true;
+    if (auto ce = e.isCallExp())
+    {
+        auto fd = ce.f;
+        auto id = fd ? fd.ident : null;
+        // opXXX are reserved identifiers, so it's ok to match too much here
+        if (id && id.toString().startsWith("op") && id.toString().endsWith("Assign"))
+            return true;
+    }
+    return false;
+}
+
+/**
+ * Returns true if `s` starts with `prefix`, false otherwise.
+ */
+private bool startsWith(const(char)[] s, const(char)[] prefix)
+{
+    if (s.length < prefix.length)
+        return false;
+    return s[0 .. prefix.length] == prefix;
+}
+
+/**
+ * Returns true if `s` ends with `suffix`, false otherwise.
+ */
+private bool endsWith(const(char)[] s, const(char)[] suffix)
+{
+    if (s.length < suffix.length)
+        return false;
+    return s[$ - suffix.length .. $] == suffix;
+}
+
+/**
+ * Returns true if the given expression is an increment (++) or decrement (--).
+ */
+private bool isIncrementOrDecrement(Expression e)
+{
+    import dmd.dtemplate : isExpression;
+    import dmd.globals : Loc;
+    import dmd.tokens : EXP;
+
+    if (e.op == EXP.plusPlus
+        || e.op == EXP.minusMinus
+        || e.op == EXP.prePlusPlus
+        || e.op == EXP.preMinusMinus)
+        return true;
+    if (auto call = e.isCallExp())
+    {
+        // Check for overloaded preincrement
+        // e.g., a.opUnary!"++"
+        auto fd = call.f;
+        auto id = fd ? fd.ident : null;
+        if (id && id.toString() == "opUnary")
+        {
+            auto ti = fd.parent ? fd.parent.isTemplateInstance() : null;
+            auto tiargs = ti ? ti.tiargs : null;
+            if (tiargs && tiargs.length >= 1 && (*tiargs)[0].isExpression())
+            {
+                auto op = (cast(Expression) (*tiargs)[0]).isStringExp();
+                scope plusPlus = new StringExp(Loc.initial, "++");
+                scope minusMinus = new StringExp(Loc.initial, "--");
+                if (op && (op.compare(plusPlus) == 0 || op.compare(minusMinus) == 0))
+                    return true;
+            }
+        }
+    }
+    else if (auto comma = e.isCommaExp())
+    {
+        // Check for overloaded postincrement
+        // e.g., (auto tmp = a, ++a, tmp)
+        auto left = comma.e1 ? comma.e1.isCommaExp() : null;
+        auto middle = left ? left.e2 : null;
+        if (middle && isIncrementOrDecrement(middle))
+            return true;
+    }
+    return false;
+}
+
+/**
+ * Returns true if the given symbol has the @mustUseAttribute.
+ */
+private bool hasMustUseAttribute(Dsymbol sym, Scope* sc)
+{
+    import dmd.attrib : foreachUda;
+
+    bool result = false;
```

Comment:
Redundant to the return value of `foreachUda`

Reply:
The return value of `foreachUda` indicates whether or not the loop exited early. In this case, that return value happens to be the same as `result`, but (a) there's no inherent reason that *has* to be true, so it could change in a refactor, and (b) the explicit `result` makes the code easier to understand in isolation (esp. if one hasn't read `foreachUda`'s documentation recently).

Reply:
(A) would imply that one can rely on the return value of any function. Call sites need to be revised when changing the behaviour any function.

(B) ok